OPAL/br/src/main/scala/org/opalj/br/fpcf/properties/string_definition/StringTree.scala

```diff
@@ -0,0 +1,385 @@
+        val unique = ListBuffer[StringTreeElement]()
+        children.foreach {
+            case next @ StringTreeConst(sci) ⇒
+                if (!seen.contains(sci)) {
+                    seen += (sci → true)
+                    unique.append(next)
+                }
+            case loop: StringTreeRepetition ⇒ unique.append(loop)
+            case concat: StringTreeConcat   ⇒ unique.append(concat)
+            case or: StringTreeOr           ⇒ unique.append(or)
+            case cond: StringTreeCond       ⇒ unique.append(cond)
+        }
+        unique
+    }
+
+    /**
+     * Accumulator function for simplifying a tree.
+     */
+    private def simplifyAcc(subtree: StringTree): StringTree = {
+        subtree match {
+            case StringTreeOr(cs) ⇒
+                cs.foreach {
+                    case nextC @ StringTreeOr(subChildren) ⇒
+                        simplifyAcc(nextC)
+                        var insertIndex = subtree.children.indexOf(nextC)
+                        subChildren.foreach { next ⇒
+                            subtree.children.insert(insertIndex, next)
+                            insertIndex += 1
+                        }
+                        subtree.children.-=(nextC)
+                    case _ ⇒
+                }
+                val unique = removeDuplicateTreeValues(cs)
+                subtree.children.clear()
+                subtree.children.appendAll(unique)
+                subtree
+            case stc: StringTreeCond ⇒
+                // If the child of a StringTreeCond is a StringTreeRepetition, replace the
+                // StringTreeCond by the StringTreeRepetition element (otherwise, regular
+                // expressions like ((s)*)+ will follow which is equivalent to (s)*).
+                if (stc.children.nonEmpty && stc.children.head.isInstanceOf[StringTreeRepetition]) {
+                    stc.children.head
+                } else {
+                    stc
+                }
+            // Remaining cases are trivial
+            case str: StringTreeRepetition ⇒ StringTreeRepetition(simplifyAcc(str.child))
+            case stc: StringTreeConcat     ⇒ StringTreeConcat(stc.children.map(simplifyAcc))
+            case stc: StringTreeConst      ⇒ stc
+        }
+    }
+
+    /**
+     * Accumulator function for grouping repetition elements.
+     */
+    private def groupRepetitionElementsAcc(subtree: StringTree): StringTree = {
+        /**
+         * Function for processing [[StringTreeOr]] or [[StringTreeConcat]] elements as these cases
+         * are equal (except for distinguishing the object to return). Thus, make sure that only
+         * instance of these classes are passed. Otherwise, an exception will be thrown!
+         */
+        def processConcatOrOrCase(subtree: StringTree): StringTree = {
+            if (!subtree.isInstanceOf[StringTreeOr] && !subtree.isInstanceOf[StringTreeConcat]) {
```

Comment:
There are a couple of isInstanceOf/asInstanceOf that could be replaced by pattern matching.